diff --git a/Makefile b/Makefile
index e98e3ca..128de80 100644
--- a/Makefile
+++ b/Makefile
@@ -101,13 +101,13 @@ EXTRA_CFLAGS += -DCONFIG_RTW_ANDROID=$(CONFIG_RTW_ANDROID)
 endif
 
 ########################## Debug ###########################
-CONFIG_RTW_DEBUG = y
+CONFIG_RTW_DEBUG = n
 # default log level is _DRV_INFO_ = 4,
 # please refer to "How_to_set_driver_debug_log_level.doc" to set the available level.
 CONFIG_RTW_LOG_LEVEL = 3
 
 # enable /proc/net/rtlxxxx/ debug interfaces
-CONFIG_PROC_DEBUG = y
+CONFIG_PROC_DEBUG = n
 
 ######################## Wake On Lan ##########################
 CONFIG_WOWLAN = n
@@ -137,7 +137,8 @@ CONFIG_LAYER2_ROAMING = y
 #bit0: ROAM_ON_EXPIRED, #bit1: ROAM_ON_RESUME, #bit2: ROAM_ACTIVE
 CONFIG_ROAMING_FLAG = 0x3
 ###################### Platform Related #######################
-CONFIG_PLATFORM_I386_PC = y
+CONFIG_PLATFORM_OPENWRT = y
+CONFIG_PLATFORM_I386_PC = n
 CONFIG_PLATFORM_ANDROID_X86 = n
 CONFIG_PLATFORM_ANDROID_INTEL_X86 = n
 CONFIG_PLATFORM_JB_X86 = n
@@ -1336,6 +1337,34 @@ ifeq ($(CONFIG_RTW_IOCTL_SET_COUNTRY), y)
 EXTRA_CFLAGS += -DCONFIG_RTW_IOCTL_SET_COUNTRY
 endif
 
+ifeq ($(CONFIG_PLATFORM_OPENWRT), y)
+include $(TOPDIR)/rules.mk
+include $(INCLUDE_DIR)/kernel.mk
+
+EXTRA_CFLAGS += -DCONFIG_LITTLE_ENDIAN \
+		-DCONFIG_IOCTL_CFG80211 \
+		-DRTW_USE_CFG80211_STA_EVENT \
+		-DRTW_IPS_MODE=0 \
+		-DRTW_LPS_MODE=0 \
+		-DRTW_WOW_LPS_MODE=0 \
+		-DCONFIG_TDLS \
+		-DCONFIG_RTW_MESH \
+		-DCONFIG_RTW_MULTI_AP \
+		-DCONFIG_P2P \
+		-DCONFIG_FORCE_SW_CHANNEL_PLAN \
+		-DCONFIG_CONCURRENT_MODE \
+		-DCONFIG_IFACE_NUMBER=4 \
+		-DCONFIG_QOS_OPTIMIZATION \
+		-DRTW_REDUCE_SCAN_SWITCH_CH_TIME \
+		-Wno-error
+
+USER_MODULE_NAME := rtl$(MODULE_NAME)
+ARCH="$(LINUX_KARCH)"
+CROSS_COMPILE="$(TARGET_CROSS)"
+KSRC="$(LINUX_DIR)"
+KVER="$(LINUX_VERSION)"
+endif
+
 ifeq ($(CONFIG_PLATFORM_I386_PC), y)
 EXTRA_CFLAGS += -DCONFIG_LITTLE_ENDIAN
 EXTRA_CFLAGS += -DCONFIG_IOCTL_CFG80211 -DRTW_USE_CFG80211_STA_EVENT
diff --git a/core/mesh/rtw_mesh.c b/core/mesh/rtw_mesh.c
index fad46e8..5da72f4 100644
--- a/core/mesh/rtw_mesh.c
+++ b/core/mesh/rtw_mesh.c
@@ -3022,6 +3022,7 @@ static void rtw_mpath_tx_queue_flush(_adapter *adapter)
 
 #ifdef PLATFORM_LINUX /* 3.10 ~ 4.13 checked */
 #if defined(CONFIG_SLUB)
+#include <../mm/slab.h>
 #include <linux/slub_def.h>
 #elif defined(CONFIG_SLAB)
 #include <linux/slab_def.h>
diff --git a/os_dep/linux/ioctl_cfg80211.c b/os_dep/linux/ioctl_cfg80211.c
index f7b59a8..ac1e0b2 100644
--- a/os_dep/linux/ioctl_cfg80211.c
+++ b/os_dep/linux/ioctl_cfg80211.c
@@ -8413,7 +8413,7 @@
 	return ret;
 }
 
-#if (LINUX_VERSION_CODE < KERNEL_VERSION(5, 8, 0))
+#if (LINUX_VERSION_CODE < KERNEL_VERSION(5, 8, 0)) && ! defined (BUILD_OPENWRT)
 static void cfg80211_rtw_mgmt_frame_register(struct wiphy *wiphy,
 #if (LINUX_VERSION_CODE >= KERNEL_VERSION(3, 6, 0))
 	struct wireless_dev *wdev,
@@ -10595,7 +10595,7 @@
 
 #if (LINUX_VERSION_CODE >= KERNEL_VERSION(2, 6, 37)) || defined(COMPAT_KERNEL_RELEASE)
 	.mgmt_tx = cfg80211_rtw_mgmt_tx,
-#if (LINUX_VERSION_CODE < KERNEL_VERSION(5, 8, 0))
+#if (LINUX_VERSION_CODE < KERNEL_VERSION(5, 8, 0)) && ! defined (BUILD_OPENWRT)
 	.mgmt_frame_register = cfg80211_rtw_mgmt_frame_register,
 #else
 	.update_mgmt_frame_registrations = cfg80211_rtw_update_mgmt_frame_register,
